Convert run_tests.sh to a Python script (#498)

agundy opened the pull request:
Replaces run_tests.sh with run_tests.py for long term maintablility.

whimboo commented:
Hi @agundy. I would like to ask if you can find the time to finish the last review comments anytime soon. If you have questions about specific requests please let me know. Thanks!

agundy commented on run_tests.py:
```diff
@@ -0,0 +1,110 @@
+
+        pip('install', 'selenium')
+
+        python('test/configuration/save_config.py')
```
Should I check for Windows like they do? Or is it assumed to be running on Linux?

whimboo replied:
We do not officially support Windows for now, but it wouldn't be a problem to add those checks. In case you do not have Windows, please just ignore that.

agundy commented:
I definitely can. Sorry about the absence. I added a question about activating the environment. I think i fixed everything else up. Once I activate it correctly I will push the changes.

agundy commented:
@whimboo Did you see my last commit? I think I fixed everything that you were looking for. I pushed it 10 days ago.

whimboo commented:
@agundy sorry but I missed that commit because github does not automatically inform me via an email about that. Please make sure to add at least a small comment.

Currently this PR cannot be merged. Would you mind to rebase the code against master?

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,117 @@
+import pip
+import urllib2
+import tarfile
+import shutil
```
Please sort the imports alphabetically, and move the one where you use `from` to its own block at the end.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,117 @@
+    """Main Function that handles all the tests."""
+    try:
+        print "Checking Patches"
+        check_call('./test/check_patches.sh')
```
Please reduce the size of code which has been put into this try/except. The message you show in except only relates to activating the environment. So if necessary use multiple try/except or finally blocks.

This specific `check_call` here can perfectly be moved out of the try block. It throws an exception, but there is nothing we would have to clean-up.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,117 @@
+                user_input = raw_input("Would you like to recreate it? ")
+                if user_input[0].lower() == "y":
+                    print "Running setup:"
+                    shutil.rmtree(DIR_JENKINS_ENV, True)
```
setup.sh cleans up an existent environment folder automatically. So you can remove this line.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,117 @@
+            while True:
+                user_input = raw_input("Would you like to recreate it? ")
+                if user_input[0].lower() == "y":
+                    print "Running setup:"
```
nit: you can remove the trailing colon.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,117 @@
+                    print "Please answer yes or no."
+        else:
+            print "Running setup"
+            check_call(["./setup.sh", DIR_JENKINS_ENV])
```
All the above also doesn't need any clean-up code, so you can move it out of the try block.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,117 @@
+            check_call(["./setup.sh", DIR_JENKINS_ENV])
+
+        print "Starting Jenkins"
+        Popen(['./start.py', '>', 'jenkins.out'])
```
So this is routing the stdout to the log file. It would be better to specify that as a separate kwarg in the Popen() call. See here for an example: http://stackoverflow.com/a/3190840

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,117 @@
+
+        time.sleep(60)
+
+        if os.path.exists(DIR_TEST_ENV):
```
Please also keep the comment from the other file to make it better understandable.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,117 @@
+        if os.path.exists(DIR_TEST_ENV):
+            print "Using virtual environment in ", DIR_TEST_ENV
+        else:
+            tar_filename = 'virtualenv-%s.tar.gz' %VERSION_VIRTUALENV
```
nit: missing blank after the %.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,117 @@
+        else:
+            tar_filename = 'virtualenv-%s.tar.gz' %VERSION_VIRTUALENV
+            print 'Creating a virtual environment (version %s) in %s' % (VERSION_VIRTUALENV, DIR_TEST_ENV)
+            url = 'https://pypi.python.org/packages/source/v/virtualenv/%s' %tar_filename
```
nit: same missing blank here.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,117 @@
+            tar = tarfile.open(name=tar_filename)
+            tar.extractall(path='.')
+            check_call(
+                ["virtualenv-%s/virtualenv.py" %VERSION_VIRTUALENV , DIR_TEST_ENV])
```
Move this line up and if necessary break before `DIR_TEST_ENV`. Also add a blank after %.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,117 @@
+        print 'Activating virtual environment'
+        # Assumes Linux Platform
+        activate_this_file = os.path.join(DIR_TEST_ENV, 'bin',
+                                                    'activate_this.py')
```
Please indent with DIR_TEST_ENV.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,117 @@
+            check_call(['virtualenv', '--no-site-packages', DIR_ENV])
+
+        execfile(activate_this_file, dict(__file__=activate_this_file))
+        print "Virtual environment activated successfully."
```
Would it work to move all the environment creation code before we start jenkins in line 74? If something fails here we wouldn't have to wait 60s.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,117 @@
+        kill_jenkins()
+        sys.exit(1)
+    time.sleep(60)
+    kill_jenkins()
```
You can use `finally` to ensure we quit the start.py script. Therefor keep the process id from the Popen() call, which let you then quit the process.

whimboo commented:
@agundy I would like to know if you will have the time to continue your work on this PR. We got very far and it shouldn't take that long anymore to finish up the remaining nits. Please let me know. Thanks.

agundy commented:
@whimboo I can definitely find time to work on this. The semester kind of got on top of me and I was a little overwhelmed with all the feedback you had given me at the time. I will look over your comments tomorrow and update some of the code. I would love your feedback then.

whimboo commented:
That sounds great. To reduce the amount of feedback it has been proven to work together on some parts while staying on IRC for conversations. If that would be possible for you please let me know.

agundy commented:
I can do IRC. Where can I find you on IRC?

whimboo commented:
That's great! Just keep in mind that I live in Europe and might not be around late in the evening each day. But when I'm around you can find me as whimboo in the #automation channel. If you don't have a local IRC client you can simply use the web client: https://kiwiirc.com/client/irc.mozilla.org/#automation. I'm looking forward to meeting you there.

agundy commented:
Hi whimboo. I pushed updates to this. I can hop on IRC and review the changes if you have time.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,116 @@
+
+
+def python(*args):
+    pypath = [os.path.join(DIR_TEST_ENV, 'bin', 'python')]
```
Doesn't `sys.executable` return us the right Python executable even when we have activated the environment? Could you please check this?

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,116 @@
+
+
+def pip(*args):
+    pippath = [os.path.join(DIR_TEST_ENV, 'bin', 'pip')]
```
When we have activated the environment ´pip´ should be in the path and you won't need the full path. Constructing that one would make this code fail on Windows, because there is no ´bin´ but a `Scripts` folder.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,116 @@
+    try:
+        print "Starting Jenkins"
+        log = open('jenkins.out', 'w')
+        error = open('jenkins.error', 'w')
```
If we don't need those files lets not add them to the Popen call. It would be better to see the output in the console which is done by default.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,116 @@
+
+        pip('install', 'selenium')
+
+        time.sleep(60)
```
Can we also execute the ´start.py´ script here? It would move the invocation of Jenkins closer to the code below. We might loose a bit of time right now, but with a followup issue we could check the ready state of Jenkins. When doing all that you would only have to use try/except around the code which interacts with start.py, and ensures to clean-up any running instance at the end.

whimboo commented on run_tests.py:
```diff
@@ -0,0 +1,116 @@
+        print "Exiting"
+        kill_jenkins()
+        sys.exit(1)
+    kill_jenkins()
```
Instead of calling `kill_jenkins()` twice you can use the `finally` statement, which will always be executed.

whimboo commented:
Good enhancements Aaron! Lets chat on IRC if the remaining items are not that clear for you. I feel that we are close to being ready to land the patch. Thanks!

whimboo commented:
Hi @agundy. I would like to ask if you had the time to look at this PR again and if we can have a chat on IRC as proposed. Thanks.

whimboo commented:
Looks like Aaron is no longer working on this PR. So I will put it back into the general bucket and maybe someone else can finish off the remaining work. Aaron, feel free to reopen in case you still have time and interest to finish it off. Thanks.
